Introduce a new entity func ReadWithOptions.

Spiral90210 opened the pull request:
This provides a way to customise the behaviour of Read, starting with the max permissible size headers can be. Resolves #149

brunnre8 commented on entity.go:
```diff
@@ -101,15 +101,29 @@ func (lr *limitedReader) Read(p []byte) (int, error) {
+
+// checkValues will modify the ReadOptions struct setting defaults or special
+// values. Must be called before the ReadOptions struct is used.
+func (o *ReadOptions) checkValues() {
```
checkValues seems like an awkward name, considering that you aren't checking values (meaning returning a bool or error or some such) but modify them.

Spiral90210 replied:
I agree, it wasn't easy to name. I initially had `setDefaults()`, but thought that might be confused with a constructor type func. Also considered `validate()` but that has the same issues as `checkValues()`. `setDefaults()` is probably the best fit, but I'm open to other names or ways to approach this.

Spiral90210 replied:
Perhaps another approach would be to take a pointer receiver like you suggested, but have a func `readerOptionsFromUserOpts(userOpts *ReaderOptions) (*ReaderOptions)` (or something else, I'm terrible at short names), which would just make a new instance that is guaranteed valid?

emersion replied:
Yeah, sounds good. `ReaderOptions.withDefaults` may be a reasonable name?

brunnre8 commented on entity_test.go:
```diff
@@ -162,6 +163,16 @@ func TestRead_tooBig(t *testing.T) {
 	}
 }
 
+func TestReadWithOptions_canSetMaxHeaderBytes(t *testing.T) {
```
You aren't checking any of the default value patching is working, only the manual override.
At the very least you should check for -1, 0 and some manual value in a test and validate that the option struct contains what you think it does.

Spiral90210 replied:
The existing Read func will check that 0 (int default) is mapped to something non 0 (otherwise it would fail). You're right I did not check -1, I'll add some more tests this evening.

Spiral90210 replied:
Now done.

brunnre8 reviewed (commented):
some things I've noticed

emersion commented on entity.go:
```diff
@@ -101,15 +101,29 @@ func (lr *limitedReader) Read(p []byte) (int, error) {
+
+// ReadWithOptions see Read, but allows overriding some parameters with
+// ReadOptions.
+func ReadWithOptions(r io.Reader, opts ReadOptions) (*Entity, error) {
```
In general options are a pointer to a struct (see other examples in the stdlib).

Spiral90210 replied:
This was actually deliberate - because the (currently named) `checkValues()` func mutates the struct, I didn't want to do that to a struct that I didn't own, it seems unexpected. By taking a value here, we get a copy that we are free to mutate.

Spiral90210 replied:
I've changed this to a pointer receiver in the latest commit, resolving this convo.

Spiral90210 commented:
I've updated ReadWithOptions to take a value receiver, and changed checkValues() -> withDefaults() which produces a pointer to a new value. There's a test to ensure no mutation of the original options, and much better test coverage of different option scenarios. Thanks for the quick feedback, sorry I didn't get it done last night as I'd said.

emersion commented:
It would be nice to allow to pass nil to get all default options.

Spiral90210 commented:
> It would be nice to allow to pass nil to get all default options.

Even better to not crash 😆

emersion commented:
Thanks!
